### PR TITLE
[Bagel]: Support TP

### PR DESCRIPTION
**Pull request by princepride**
## Purpose
https://github.com/vllm-project/vllm-omni/issues/1253 Let Bagel support TP

## Test Plan
```
from PIL import Image
from vllm_omni.entrypoints.omni_diffusion import OmniDiffusion
from vllm_omni.inputs.data import OmniDiffusionSamplingParams, OmniPromptType
import time

def main():
    image = Image.open("women.jpg")
    pipeline = OmniDiffusion(
        model="../models/BAGEL-7B-MoT",
        parallel_config={
            "tensor_parallel_size": 2
        }
    )
    prompts = {
        "prompt": "Let the woman wear a blue dress",
        "multi_modal_data": {"image": image},
    }
    
    result = pipeline.generate(
        prompts,
        OmniDiffusionSamplingParams(
            seed=52
        )
    )
    result[0].images[0].save("bagel_i2i_output.png")

if __name__ == "__main__":
    main()
```
### Model output:
<details>
image
</details>

### `TP = 1` memory usage:
<details>
image
</details>

### `TP = 2` memory usage:
<details>
image

</details>

### CODE REVIEW

**Comment by princepride** [comment]
@hsliuustc0106 PTAL

**Comment by hsliuustc0106** [comment]
update examples as well

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py [hunk 1]
```diff
@@ -66,6 +71,92 @@ def forward(self, x):
...
+        inv_freq_expanded = self.inv_freq[None, :, None].float().expand(position_ids.shape[0], -1, 1)
+        position_ids_expanded = position_ids[:, None, :].float()
+        freqs = (inv_freq_expanded @ position_ids_expanded).transpose(1, 2)
+        emb = torch.cat((freqs, freqs), dim=-1)
+        cos = emb.cos() * self.attention_scaling
+        sin = emb.sin() * self.attention_scaling
+        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)
+
+
+class BagelMLP(nn.Module):
```
Comment (Gaohan123) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: Why do we need a special BagelMLP. What is the difference from general one?

**Comment by princepride** [comment]
@hsliuustc0106 Ready to merge.

File: vllm_omni/diffusion/models/bagel/bagel_transformer.py
Comment (ZJY0516) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py]: I think these can be parallelized as well.

Comment (ZJY0516) [inline vllm_omni/diffusion/models/bagel/bagel_transformer.py:59]: same

**Comment by ZJY0516** [comment]
The gpu mem utilization indicates that some linear layers are not splited.

**Comment by princepride** [comment]
@ZJY0516 Mainly copied from: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/qwen2.py: Qwen2Attention, because bagel used a different rope and add a lot of qkv_moe module so I didn't inherit it, I also update the memory usage of the new version of this model.

File: examples/online_serving/bagel/README.md [hunk 2]
```diff
@@ -22,12 +22,29 @@ cd /workspace/vllm-omni/examples/online_serving/bagel
...
+      tensor_parallel_size: 2
+      ...
+    runtime:
+      devices: "0,1"
+```
+
+2. **Launch Server**:
+```bash
+vllm serve ByteDance-Seed/BAGEL-7B-MoT --omni --port 8091 --stage-configs-path /path/to/your/custom_bagel.yaml
+```
```
Comment (wtomin) [inline examples/online_serving/bagel/README.md:46]: Is TP online serving supported by CLI argument like `--tp 2`?

Comment (princepride) [reply examples/online_serving/bagel/README.md:46]: I am afraid not 😂, CLI argument tp can't overwrite the yaml config:
```
(APIServer pid=1640082) INFO 02-10 01:16:51 [utils.py:261] non-default args: {'model_tag': 'ByteDance-Seed/BAGEL-7B-MoT', 'port': 8091, 'model': 'ByteDance-Seed/BAGEL-7B-MoT', 'tensor_parallel_size': 2}
(APIServer pid=1640082) INFO 02-10 01:16:51 [omni.py:117] Initializing stages for model: ByteDance-Seed/BAGEL-7B-MoT
(APIServer pid=1640082) The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
(APIServer pid=1640082) The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
(APIServer pid=1640082) The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
(APIServer pid=1640082) The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
(APIServer pid=1640082) INFO 02-10 01:16:51 [initialization.py:197] Auto-configuring SharedMemoryConnector for edge ('0', '1')
(APIServer pid=1640082) INFO 02-10 01:16:51 [initialization.py:234] Loaded OmniTransferConfig with 1 connector configurations
(APIServer pid=1640082) INFO 02-10 01:16:51 [factory.py:46] Created connector: SharedMemoryConnector
(APIServer pid=1640082) INFO 02-10 01:16:51 [initialization.py:60] Created connector for 0 -> 1: SharedMemoryConnector
(APIServer pid=1640082) INFO 02-10 01:16:51 [omni_stage.py:239] [OmniStage] stage_config: {'stage_id': 0, 'stage_type': 'llm', 'runtime': {'devices': '0', 'max_batch_size': 1}, 'engine_args': {'model_stage': 'thinker', 'model_arch': 'BagelForConditionalGeneration', 'worker_type': 'ar', 'scheduler_cls': 'vllm_omni.core.sched.omni_ar_scheduler.OmniARScheduler', 'gpu_memory_utilization': 0.35, 'enforce_eager': True, 'trust_remote_code': True, 'engine_output_type': 'text', 'distributed_executor_backend': 'mp', 'enable_prefix_caching': False, 'max_num_batched_tokens': 32768, 'tensor_parallel_size': 1, 'omni_kv_config': {'need_send_cache': True, 'kv_transfer_criteria': {'type': 'prefill_finished'}}, 'max_num_seqs': 1, 'async_chunk': False}, 'final_output': True, 'final_output_type': 'text', 'is_comprehension': True, 'default_sampling_params': {'temperature': 0.4, 'top_p': 0.9, 'top_k': 1, 'max_tokens': 2048, 'seed': 52, 'detokenize': True, 'repetition_penalty': 1.05}}
(APIServer pid=1640082) INFO 02-10 01:16:51 [omni_stage.py:239] [OmniStage] stage_config: {'stage_id': 1, 'stage_type': 'diffusion', 'runtime': {'devices': '0', 'max_batch_size': 1}, 'engine_args': {'model_stage': 'dit', 'gpu_memory_utilization': 0.55, 'enforce_eager': True, 'trust_remote_code': True, 'engine_output_type': 'image', 'distributed_executor_backend': 'mp', 'enable_prefix_caching': False, 'max_num_batched_tokens': 32768, 'tensor_parallel_size': 1, 'omni_kv_config': {'need_recv_cache': True}}, 'engine_input_source': [0], 'final_output': True, 'final_output_type': 'image', 'is_comprehension': False, 'default_sampling_params': {'seed': 52}}
```

Comment (princepride) [reply examples/online_serving/bagel/README.md:46]: @lishunyang12 Can we overwrite it in the future?
